src/Numerics/LinearSolvers/BatchedGeneralizedMinimalResidualSolver.jl

```diff
@@ -0,0 +1,824 @@
+    # degrees of freedom per physical mesh column
+    reshaping_tup = (Nq..., num_states, nvertelem, nhorzelem)
+
+    if independent_states
+        m = Nglp * nvertelem
+        n = (Nglp^(dim - 1)) * nhorzelem * num_states
+    else
+        m = Nglp * nvertelem * num_states
+        n = (Nglp^(dim - 1)) * nhorzelem
+    end
+
+    if max_subspace_size === nothing
+        max_subspace_size = m
+    end
+
+    # permute_tuple_f = (3, 5, 4, 2, 1, 6)
+    # Now we need to determine an appropriate permutation
+    # of the MPIStateArray to perform column-wise strides.
+    # Total size of the permute tuple
+    permute_size = length(reshaping_tup)
+
+    # Index associated with number of GL points
+    # in the 'vertical' direction
+    nql = length(Nq)
+    # Want: (index associated with GL pts in vertical direction,
+    #        index associated with the stack size of the column,
+    #        index associated with number of states)
+    # FIXME: Better way to do this?
+    column_strides = (nql, nql + 2, nql - 1)
+    diff = Tuple(setdiff(Set([i for i in 1:permute_size]), Set(column_strides)))
+    permute_tuple_f = (column_strides..., diff...)
+
+    return BatchedGeneralizedMinimalResidual(
+        Q;
+        m = m,
+        n = n,
+        subspace_size = max_subspace_size,
+        atol = atol,
+        rtol = rtol,
+        ArrayType = ArrayType,
+        reshape_tuple_f = reshaping_tup,
+        permute_tuple_f = permute_tuple_f,
+    )
+end
+
+# initialize function (1)
+function LS.initialize!(
+    linearoperator!,
+    Q,
+    Qrhs,
+    solver::BatchedGeneralizedMinimalResidual,
+    args...,
+)
+    # body of initialize function in abstract iterative solver
+    return false, zero(eltype(Q))
+end
+
+# iteration function (2)
+function LS.doiteration!(
+    linearoperator!,
+    Q,
+    Qrhs,
+    gmres::BatchedGeneralizedMinimalResidual,
+    threshold,
+    args...,
+)
+    # initialize gmres.x
+    convert_structure!(gmres.x, Q, gmres.reshape_tuple_f, gmres.permute_tuple_f)
+    # apply linear operator to construct residual
+    r_vector = copy(Q)
+    linearoperator!(r_vector, Q, args...)
+    @. r_vector = Qrhs - r_vector
+    # The following ar and rr are technically not correct in general cases
+    ar = norm(r_vector)
+    rr = norm(r_vector) / norm(Qrhs)
+    # check if the initial guess is fantastic
+    if (ar < gmres.atol) || (rr < gmres.rtol)
+        return true, 0, ar
+    end
+    # initialize gmres.b
+    convert_structure!(
+        gmres.b,
+        r_vector,
+        gmres.reshape_tuple_f,
+        gmres.permute_tuple_f,
+    )
+    # apply linear operator to construct second krylov vector
+    linearoperator!(Q, r_vector, args...)
+    # initialize gmres.sol
+    convert_structure!(
+        gmres.sol,
+        Q,
+        gmres.reshape_tuple_f,
+        gmres.permute_tuple_f,
+    )
+    # initialize the rest of gmres
+    event = initialize_gmres!(gmres)
+    wait(event)
```

Comment:
Can we get away with a device wait here?  Either `wait(CPU(), event)` or `wait(GPU(), event)`.